### PR TITLE
Reflection types cleanup

### PR DESCRIPTION
**Pull request by kukulich**

### CODE REVIEW

File: src/Reflection/ReflectionIntersectionType.php [hunk 1]
```diff
@@ -18,15 +18,16 @@ class ReflectionIntersectionType extends ReflectionType
...
+        private ReflectionParameter|ReflectionMethod|ReflectionFunction|ReflectionEnum|ReflectionProperty $owner,
         IntersectionType $type,
     ) {
-        parent::__construct($reflector, $owner);
-
-        $this->types = array_values(array_map(static function (Node\Identifier|Node\Name $type) use ($reflector, $owner): ReflectionNamedType {
-            $type = ReflectionType::createFromNode($reflector, $owner, $type);
+        $this->types = array_values(array_map(function (Node\Identifier|Node\Name $type): ReflectionNamedType {
+            $type = ReflectionType::createFromNode($this->reflector, $this->owner, $type);
             assert($type instanceof ReflectionNamedType);
```
Comment (Ocramius) [inline src/Reflection/ReflectionIntersectionType.php:29]: This will change with PHP 8.2

File: src/Reflection/ReflectionUnionType.php [hunk 2]
```diff
@@ -18,15 +18,16 @@ class ReflectionUnionType extends ReflectionType
...
+        private ReflectionParameter|ReflectionMethod|ReflectionFunction|ReflectionEnum|ReflectionProperty $owner,
         UnionType $type,
     ) {
-        parent::__construct($reflector, $owner);
-
-        $this->types = array_values(array_map(static function (Node\Identifier|Node\Name $type) use ($reflector, $owner): ReflectionNamedType {
-            $type = ReflectionType::createFromNode($reflector, $owner, $type);
+        $this->types = array_values(array_map(function (Node\Identifier|Node\Name $type): ReflectionNamedType {
+            $type = ReflectionType::createFromNode($this->reflector, $this->owner, $type);
             assert($type instanceof ReflectionNamedType);
```
Comment (Ocramius) [inline src/Reflection/ReflectionUnionType.php:29]: This will change with PHP 8.2

File: src/Reflection/ReflectionType.php [hunk 3]
```diff
@@ -73,14 +67,6 @@ public static function createFromNode(
         return new ReflectionUnionType($reflector, $owner, new UnionType($types));
     }
 
-    /**
-     * @internal
-     */
-    public function getOwner(): ReflectionParameter|ReflectionMethod|ReflectionFunction|ReflectionEnum|ReflectionProperty
```
Comment (Ocramius) [inline src/Reflection/ReflectionType.php:79]: Hmm, removing this means no longer being able to run some checks in `roave/backwards-compatibility-check` 🤔

Comment (kukulich) [reply src/Reflection/ReflectionType.php:79]: I can also remove the `@internal` flag :)

Comment (Ocramius) [reply src/Reflection/ReflectionType.php:79]: I'm not saying we shouldn't do it, I'm saying that we need some sort of approach to how to deal with it downstream :D 

Practically, if we move reflection stuff to be more and more immutable, we should totally do this.

Comment (kukulich) [reply src/Reflection/ReflectionType.php:79]: Do you really use this method in `roave/backwards-compatibility-check`? I'm not able to find it. Didn't you mean https://github.com/Roave/BetterReflection/blob/5.7.x/src/Reflection/ReflectionNamedType.php#L65 ?

Comment (kukulich) [reply src/Reflection/ReflectionType.php:79]: History of the method: https://github.com/Roave/BetterReflection/pull/975

Comment (Ocramius) [reply src/Reflection/ReflectionType.php:79]: Hmm, perhaps we can try this with the other lib, as I may already have worked around it 🤷

Comment (Ocramius) [reply src/Reflection/ReflectionType.php:79]: Checked again: LGTM, let's roll with it :)

**Review by Ocramius — APPROVED** [review]
:ship: thanks @kukulich!

**Pull request merged**